pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/merge_rollup/MergeRollupTaskGenerator.java

```diff
@@ -0,0 +1,471 @@
+        // 2. All segments are merged in the merge window and we have loop through all segments, skip scheduling
+        // 3. Merge window is invalid (windowEndMs > System.currentTimeMillis() - bufferMs || windowEndMs > waterMark of
+        //    the lower mergeLevel), skip scheduling
+        for (OfflineSegmentZKMetadata preSelectedSegment : preSelectedSegments) {
+          if (windowStartMs <= preSelectedSegment.getEndTimeMs() && preSelectedSegment.getStartTimeMs() < windowEndMs) {
+            // For segments overlapping with current window, add to the result list
+            selectedSegments.add(preSelectedSegment);
+            if (!isMergedSegment(preSelectedSegment, mergeLevel)) {
+              hasUnmergedSegments = true;
+            }
+          } else if (windowEndMs <= preSelectedSegment.getStartTimeMs()) {
+            // Has gone through all overlapping segments for current window
+            if (hasUnmergedSegments) {
+              // Found unmerged segments, schedule merge task for current window
+              break;
+            } else {
+              // No unmerged segments found, clean up selected segments and bump up the merge window
+              // TODO: If there are many small merged segments, we should merge them again
+              selectedSegments.clear();
+              selectedSegments.add(preSelectedSegment);
+              if (!isMergedSegment(preSelectedSegment, mergeLevel)) {
+                hasUnmergedSegments = true;
+              }
+              windowStartMs = preSelectedSegment.getStartTimeMs() / bucketMs * bucketMs;
+              windowEndMs = windowStartMs + bucketMs;
+              if (!isValidMergeWindowEndTime(windowEndMs, bufferMs, lowerMergeLevel, mergeRollupTaskMetadata)) {
+                isValidMergeWindow = false;
+                break;
+              }
+            }
+          }
+          // windowStartMs > preSelectedSegment.getEndTimeMs()
+          // Haven't find the first overlapping segment, continue to the next segment
+        }
+
+        if (!isValidMergeWindow) {
+          LOGGER.info(
+              "Window with start: {} and end: {} of mergeLevel: {} is not a valid merge window, Skipping task generation: {}",
+              windowStartMs, windowEndMs, mergeLevel, taskType);
+          continue;
+        }
+
+        if (!hasUnmergedSegments) {
+          LOGGER.info("No unmerged segments found for mergeLevel:{} for table: {}, Skipping task generation: {}",
+              mergeLevel, offlineTableName, taskType);
+          continue;
+        }
+
+        // Update watermark in 2 cases:
+        // 1. Cold start, no watermark value in MergeRollupTaskMetadata
+        // 2. Merge window start time > current watermark
+        if (mergeRollupTaskMetadata.getWatermarkMap().get(mergeLevel) == null || windowStartMs > waterMarkMs) {
+          updateWatermarkMs(offlineTableName, mergeLevel, windowStartMs, mergeRollupTaskMetadata);
+        }
```

Comment:
If we reach here, we can always update the watermark:
```suggestion
        mergeRollupTaskMetadata.getWatermarkMap().put(mergeLevel, windowStartMs);
        LOGGER.info(...);
```